Add Luminance Toggle To Image Viewer

danieldresser-ie opened the pull request:
Not sure if this is the right approach to the icon UI, but it seems to be working.

In order to support the CPU fallback path, I had to add a Saturation node to GafferImage, but this is probably worthwhile anyway.

This is currently on top of the Focus Node PR, because I needed to edit graphics.svg again, and merging changes to that can be fairly painful.

johnhaddon reviewed (changes requested):
Thanks Daniel - I've added comments inline that I think cover the things we discussed yesterday along with a couple of further small things. As discussed yesterday we should also change the luminance icon to show the RGB channels filled solid grey. It'd be good to get this merged while the focus node stuff is ongoing, so if possible could you rebase to omit those commits?

danieldresser-ie commented on python/GafferImageUI/SaturationUI.py:
```diff
@@ -0,0 +1,69 @@
+	"description",
+	"""
+	Increases or decreases the saturation of an image.  Saturation is calculated relative
+	to a standard luminance measure using the RGB coefficients < 0.2126, 0.7152, 0.0722 >
```
Fixed and squashed.

danieldresser-ie commented on src/GafferImage/Saturation.cpp:
```diff
@@ -0,0 +1,127 @@
+		return false;
+	}
+
+	ImagePlug::GlobalScope c( Context::current() );
```
Fixed and squashed.

danieldresser-ie commented on src/GafferImage/Saturation.cpp:
```diff
@@ -0,0 +1,127 @@
+	:	ColorProcessor( name )
+{
+	storeIndexOfNextChild( g_firstPlugIndex );
+	addChild( new FloatPlug( "saturation", Gaffer::Plug::In, 1.0f ) );
```
Fixed and squashed.

danieldresser-ie commented on python/GafferImageUI/ImageViewUI.py:
```diff
@@ -709,6 +709,7 @@ def __menuDefinition( self ) :
 			( "R", 0 ),
 			( "G", 1 ),
 			( "B", 2 ),
+			( "Luminance", -2 ),
```
Fixed and squashed.

danieldresser-ie commented on src/GafferImageUI/ImageGadget.cpp:
```diff
@@ -505,13 +515,13 @@ void ImageGadget::setSoloChannel( int index )
 
 	m_soloChannel = index;
-	if( m_soloChannel == -1 )
+	if( m_soloChannel < 0 )
```
I've done this switch, however this is not just making it less pessimistic ... this is adding in a dirty when switching between solo channels, which didn't previously happen.  That seems appropriate though ... was it wrong before?

danieldresser-ie commented:
I've addressed your feedback ( including the solo channel changed dirtyness conditional, which we should check the correct behaviour on ).

I've also done a couple things to look at cleaning up graphics.svg.  The first couple are very reasonable:
* the very simple contrib/scripts/cleanGraphicsSVG seems like a nice win for making simpler diffs of the svg
* making the necessary settings change to actually clean up unused swatches is nice
* and then I just manually cleaned up swatches that were dulplicated or named badly

The last one is much less clear of a win, since I don't have a way to preserve it going forward - I just manually removed the no-op swatches from graphics.svg, using vi, and having not realized quite how many of them there are.  Now that I've done it, it does make the file shorter and clearer, but there may not be any good way to enforce it in the future.  Probably still makes sense to keep the cleaner version though.

johnhaddon commented:
> I've addressed your feedback ( including the solo channel changed dirtyness conditional, which we should check the correct behaviour on ).

Thanks - I've comments on the conditional inline.

> I've also done a couple things to look at cleaning up graphics.svg.

This looks good to me, except that `layoutButton.png` is now the wrong colour. The addition of the "Gaffer Icons" title also through me for quite a while, as it introduced diffs to every single PNG file. I had to remove that until I could track down just the one broken one.

danieldresser-ie commented:
Fixed layoutButton icon and fixed comment on the conditional, and added notes to Changes.md.  I think this is ready to go, unless we want to retarget it to 0.60_maintenance?  ( It could be a tough merge if we make SVG changes before it ... do we anticipate editing icons on 0.60_maintenance? )

johnhaddon commented:
> I think this is ready to go, unless we want to retarget it to 0.60_maintenance?

I'm sure folks would appreciate having it in 0.60, but currently the addition of member data to ImageGadget makes it an ABI change.

> do we anticipate editing icons on 0.60_maintenance

I'm not sure I'd want to be against it. Maybe we should merge the SVG and Saturation node to 0.60, and then the ImageView/ImageGadget stuff to master?

danieldresser-ie commented:
I've split out #4299 as a separate commit - this will probably require a rebase once that gets merged through.

johnhaddon commented:
I've merged #4299 through to master and rebased this PR to sit on top of it. Will merge once CI passes.
